src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs

```diff
@@ -0,0 +1,561 @@
+        provider.GetRequiredService<ILogger<ALoadoutSynchronizer>>(),
+        provider.GetRequiredService<FileHashCache>(),
+        provider.GetRequiredService<IDataStore>(),
+        provider.GetRequiredService<LoadoutRegistry>(),
+        provider.GetRequiredService<DiskStateRegistry>(),
+        provider.GetRequiredService<IArchiveManager>())
+
+    {
+
+    }
+
+    #region IStandardizedLoadoutSynchronizer Implementation
+
+    /// <inheritdoc />
+    public async ValueTask<FlattenedLoadout> LoadoutToFlattenedLoadout(Loadout loadout)
+    {
+        var dict = new Dictionary<GamePath, ModFilePair>();
+
+        var sorted = (await SortMods(loadout)).ToList();
+
+        foreach (var mod in sorted)
+        {
+            if (!mod.Enabled)
+                continue;
+
+            foreach (var (_, file) in mod.Files)
+            {
+                if (file is not IToFile toFile)
+                    continue;
+
+                dict[toFile.To] = new ModFilePair { Mod = mod, File = file };
+            }
+        }
+
+        return FlattenedLoadout.Create(dict);
+    }
+
+    /// <inheritdoc />
+    public ValueTask<FileTree> FlattenedLoadoutToFileTree(FlattenedLoadout flattenedLoadout, Loadout loadout)
+    {
+        return ValueTask.FromResult(FileTree.Create(flattenedLoadout.GetAllDescendentFiles()
+            .Select(f => KeyValuePair.Create(f.Path, f.Value!.File))));
+    }
+
+
+    /// <inheritdoc />
+    public async Task<DiskState> FileTreeToDisk(FileTree fileTree, DiskState prevState, GameInstallation installation)
+    {
+        List<KeyValuePair<GamePath, HashedEntry>> toDelete = new();
+        List<KeyValuePair<AbsolutePath, IGeneratedFile>> toWrite = new();
+        List<KeyValuePair<AbsolutePath, FromArchive>> toExtract = new();
+
+        Dictionary<GamePath, DiskStateEntry> resultingItems = new();
+
+        // We'll start by scanning the game folders and comparing it to the previous state. Technically this is a
+        // three way compare between the disk state, the previous state, and the new state. However if the disk state
+        // diverges from the previous state, we'll abort, this effectively reduces the process to a two way compare.
+        foreach (var (_, location) in installation.LocationsRegister.GetTopLevelLocations())
+        {
+            await foreach (var entry in _hashCache.IndexFolderAsync(location))
+            {
+                var gamePath = installation.LocationsRegister.ToGamePath(entry.Path);
+                if (prevState.TryGetValue(gamePath, out var prevEntry))
+                {
+                    // If the file has been modified outside of the app since the last apply, we need to ingest it.
+                    if (prevEntry.Value!.Hash != entry.Hash)
+                    {
+                        HandleNeedIngest(entry);
+                        throw new UnreachableException("HandleNeedIngest should have thrown");
```

Comment:
`HandleNeedIngest` is `virtual`, so implementations can not throw. Is that method supposed to only throw or do something else?

Reply:
Anything that overrides that method is supposed to throw, as at this point the state of the apply process is not recoverable after a ingest. I'll leave it for now, but we can easily rework that code in the future if it's a problem.